README.rst
==================
googlefonts-markup
==================

.. warning:: This is alpha library

Overview
========

This is small utility to handle specs of Google Fonts in my products.

Usage
=====

.. note:: WIP

Simple case
-----------

.. code-block:: python

   >>> from googlefonts_markup import Font
   >>> noto_sans_jp = Font(family_name="Noto Sans JP")
   >>> noto_sans_jp.css_url()
   'https://fonts.googleapis.com/css2?family=Noto+Sans+JP'
   >>> noto_sans_jp.css_tag()
   '<link href="https://fonts.googleapis.com/css2?family=Noto+Sans+JP" rel="stylesheet">'

If you want only URL of CSS, use ``googlefonts_markup.shortcuts``.

.. code-block:: python

   >>> from googlefonts_markup.shortcuts import font_css_url
   >>> font_css_url("Noto Sans JP")
   'https://fonts.googleapis.com/css2?family=Noto+Sans+JP'

With italic
-----------

.. code-block:: python

   >>> from googlefonts_markup import Axis, Font
   >>> red_hat_mono = Font(family_name="Red Hat Mono", axis_list=[Axis(italic=True)])
   >>> red_hat_mono.css_url()
   'https://fonts.googleapis.com/css2?family=Red+Hat+Mono:ital,wght@1,400'

Extra attributes
----------------

.. code-block:: python

   >>> from googlefonts_markup import Font, FontSet
   >>> noto_sans_jp = Font(family_name="Noto Sans JP")
   >>> fontset = FontSet(fonts=[noto_sans_jp], display="swap")
   >>> fontset.css_url()
   'https://fonts.googleapis.com/css2?family=Noto+Sans+JP&display=swap'

Multiple fonts
--------------

.. code-block:: python

   >>> from googlefonts_markup import Font, FontSet
   >>> noto_sans_jp = Font(family_name="Noto Sans JP")
   >>> roboto_mono = Font(family_name="Roboto Mono")
   >>> fontset = FontSet(fonts=[noto_sans_jp, roboto_mono], display="swap")
   >>> fontset.css_url()
   'https://fonts.googleapis.com/css2?family=Noto+Sans+JP&family=Roboto+Mono&display=swap'

Defering on HTML
----------------

.. code-block:: python

   >>> from googlefonts_markup import Font, FontSet
   >>> font = Font(family_name="Noto Sans JP")
   >>> font.css_tag(defer=True)
   '<link href="https://fonts.googleapis.com/css2?family=Noto+Sans+JP" rel="preload" as="style" onload="this.onload=null;this.rel=\'stylesheet\'">'

Installation
============

.. code-block:: console

   pip install git+https://github.com/attakei-lab/googlefonts-markup


CHANGES.rst
==========
Changelogs
==========

v0.4.0
======

* Support ``defer`` by ``css_tag()`` method


v0.3.0
======

* Add suporting python 3.7

v0.2.0
======

* Support axis weight range

v0.1.0
======

* Initial release


tests/test_axis.py
import pytest
from pydantic import ValidationError

from googlefonts_markup.models import Axis


def test_invalid_weight():
    with pytest.raises(ValidationError):
        Axis(weight=1)


def test_italic():
    axis = Axis(italic=True)
    assert axis.value() == "1,400"


def test_multiple_weight():
    axis = Axis(weight=(100, 900))
    assert axis.value() == "0,100..900"


tests/test_font.py
from googlefonts_markup import Axis, Font


def test_name():
    font = Font(family_name="Noto Sans JP")
    assert "family=Noto+Sans+JP" in font.css_url()


def test_with_single_axis():
    axis_list = [
        Axis(weight=100),
    ]
    font = Font(family_name="Noto Sans JP", axis_list=axis_list)
    assert "family=Noto+Sans+JP:ital,wght@0,100" in font.css_url()


def test_with_multiple_axis():
    axis_list = [
        Axis(weight=100),
        Axis(italic=True, weight=100),
    ]
    font = Font(family_name="Noto Sans JP", axis_list=axis_list)
    assert "family=Noto+Sans+JP:ital,wght@0,100;1,100" in font.css_url()


tests/test_fontset.py
from googlefonts_markup import Font, FontSet


def test_single_font_only():
    font = Font(family_name="Noto Sans JP")
    fontset = FontSet(fonts=[font])
    assert "family=Noto+Sans+JP" in fontset.css_url()


def test_with_display():
    font = Font(family_name="Noto Sans JP")
    fontset = FontSet(fonts=[font], display="swap")
    assert "family=Noto+Sans+JP" in fontset.css_url()
    assert "display=swap" in fontset.css_url()
    assert "text=" not in fontset.css_url()


def test_with_text():
    font = Font(family_name="Noto Sans JP")
    fontset = FontSet(fonts=[font], text="Hello world")
    assert "family=Noto+Sans+JP" in fontset.css_url()
    assert "text=Hello+world" in fontset.css_url()


def test_multiple_fonts():
    fonts = [
        Font(family_name="Noto Sans JP"),
        Font(family_name="Roboto"),
    ]
    fontset = FontSet(fonts=fonts)
    assert "family=Noto+Sans+JP" in fontset.css_url()
    assert "family=Roboto" in fontset.css_url()


googlefonts_markup.py
from typing import List, Literal, Optional
from urllib.parse import urlencode

from pydantic import BaseModel

API_URL = "https://fonts.googleapis.com/css2"

FontDisplayValue = Literal["auto", "block", "swap", "fallback", "optional"]


class Axis(BaseModel):
    """
    Axis range definition
    """

    italic: bool = False
    """Using Italic style"""

    weight: int = 400

    def value(self) -> str:
        return f"{1 if self.italic else 0},{self.weight}"


class Font(BaseModel):
    """
    Font spec of Google Fonts
    """

    family_name: str

    axis_list: List[Axis] = []

    def spec(self):
        if len(self.axis_list) == 0:
            return self.family_name
        axis_range = "ital,wght@" + ";".join([axis.value() for axis in self.axis_list])
        return f"{self.family_name}:{axis_range}"

    def css_url(self) -> str:
        """
        Build and return URL for using itself by Google Fonts.
        """
        query = {
            "family": self.spec(),
        }
        return f"{API_URL}?{urlencode(query, safe=':@,;')}"

    def css_tag(self) -> str:
        """
        Return HTML element with ``self.css_url()``.
        """
        return f'<link href="{self.css_url()}" rel="stylesheet">'


class FontSet(BaseModel):
    """
    Fonts spec to use Google Fonts.
    """

    fonts: List[Font]

    text: Optional[str]

    display: FontDisplayValue = "auto"

    def css_url(self) -> str:
        """
        Build and return URL for using itself by Google Fonts.
        """
        query = [("family", font.spec()) for font in self.fonts]
        if self.text:
            query.append(("text", self.text))
        query.append(("display", self.display))
        return f"{API_URL}?{urlencode(query, safe=',@:;')}"

    def css_tag(self) -> str:
        """
        Return HTML element with ``self.css_url()``.
        """
        return f'<link href="{self.css_url()}" rel="stylesheet">'


googlefonts_markup/ext/sphinx.py
"""
Sphinx extension source for googlefonts-markup
"""
import logging

try:
    from sphinx.application import Sphinx
    from sphinx.config import Config
except ImportError as err:
    logging.error(f"To use {__name__}, you must install sphinx")
    raise (err)

from ..models import Font, FontSet


def convert_googlefonts(app: Sphinx, config: Config):
    font_obj = None
    if not config.googlefonts_font:
        return
    elif isinstance(config.googlefonts_font, Font):
        font_obj = config.googlefonts_font
    elif isinstance(config.googlefonts_font, str):
        font_obj = Font(family_name=config.googlefonts_font)
    elif isinstance(config.googlefonts_font, dict):
        font_obj = Font.parse_obj(config.googlefonts_font)
    else:
        raise ValueError("Invalid format googlefonts_font")
    config.googlefonts_font = font_obj


def add_fontset_url(app: Sphinx, env):
    if not app.config.googlefonts_font:
        return
    if not app.config.googlefonts_add_css_url:
        return
    fontset = FontSet(
        fonts=[app.config.googlefonts_font], display=app.config.googlefonts_display
    )
    app.add_css_file(fontset.css_url())


def setup(app: Sphinx):
    app.add_config_value("googlefonts_add_css_url", False, True)
    app.add_config_value("googlefonts_font", None, True)
    app.add_config_value("googlefonts_display", "auto", True)
    app.connect("config-inited", convert_googlefonts)
    app.connect("env-updated", add_fontset_url)


googlefonts_markup/models.py
from typing import List, Optional, Tuple, Union
from urllib.parse import urlencode

from pydantic import BaseModel, validator

# For python3.7
try:
    from typing import Literal
except ImportError:
    from typing_extensions import Literal

API_URL = "https://fonts.googleapis.com/css2"

FontDisplayValue = Literal["auto", "block", "swap", "fallback", "optional"]


class Axis(BaseModel):
    """
    Axis range definition
    """

    italic: bool = False
    """Using Italic style"""

    weight: Union[int, Tuple[int, int]] = 400

    def value(self) -> str:
        weight = (
            f"{self.weight[0]}..{self.weight[1]}"
            if isinstance(self.weight, tuple)
            else self.weight
        )
        return f"{1 if self.italic else 0},{weight}"

    @validator("weight")
    def check_weight_range(cls, v: Union[int, Tuple[int, int]]):
        if isinstance(v, int):
            assert 100 <= v <= 900, "Axis weight must be between 100 and 900"
            return v
        min_v, max_v = v
        assert 100 <= min_v <= 900, "Axis weight must be between 100 and 900"
        assert 100 <= max_v <= 900, "Axis weight must be between 100 and 900"
        assert (
            min_v < max_v
        ), "If axis weight is range, weight[0] must be less than weight[1]"
        return v


def build_css_tag(url: str, defer: bool) -> str:
    """
    Make ``link`` element from URL and donload-rule.

    :param url: Target URL
    :param defer: Flag to donload defer
    """
    attrs = {
        "href": url,
        "rel": "stylesheet",
    }
    if defer:
        attrs["rel"] = "preload"
        attrs["as"] = "style"
        attrs["onload"] = "this.onload=null;this.rel='stylesheet'"
    attr_text = " ".join(f'{k}="{v}"' for k, v in attrs.items())
    return f"<link {attr_text}>"


class Font(BaseModel):
    """
    Font spec of Google Fonts
    """

    family_name: str

    axis_list: List[Axis] = []

    def spec(self):
        if len(self.axis_list) == 0:
            return self.family_name
        axis_range = "ital,wght@" + ";".join([axis.value() for axis in self.axis_list])
        return f"{self.family_name}:{axis_range}"

    def css_url(self) -> str:
        """
        Build and return URL for using itself by Google Fonts.
        """
        query = {
            "family": self.spec(),
        }
        return f"{API_URL}?{urlencode(query, safe=':@,;')}"

    def css_tag(self, defer: bool = False) -> str:
        """
        Return HTML element with ``self.css_url()``.

        :param defer: Using defer downloading (default is False)
        """
        return build_css_tag(self.css_url(), defer)


class FontSet(BaseModel):
    """
    Fonts spec to use Google Fonts.
    """

    fonts: List[Font]

    text: Optional[str]

    display: FontDisplayValue = "auto"

    def css_url(self) -> str:
        """
        Build and return URL for using itself by Google Fonts.
        """
        query = [("family", font.spec()) for font in self.fonts]
        if self.text:
            query.append(("text", self.text))
        query.append(("display", self.display))
        return f"{API_URL}?{urlencode(query, safe=',@:;')}"

    def css_tag(self, defer: bool = False) -> str:
        """
        Return HTML element with ``self.css_url()``.

        :param defer: Using defer downloading (default is False)
        """
        return build_css_tag(self.css_url(), defer)
